docs/csharp/methods.md

```diff
@@ -1,27 +1,295 @@
+
+```cs
+var person = GetPersonalInfo("111111111");
+if (person != null)
+   Console.WriteLine("{person.FName} {person.LName}: age = {person.Age})";
+```
+
+If a method is passed an array as an argument and modifies the value of individual elements, it is not necessary for the method to return the array, although you may choose to do so for good style or functional flow of values.  This is because C# passes all reference types by value, and the value of an array reference is the pointer to the array. In the following example, changes to the contents of the `values` array that are made in the `DoubleValues` method are observable by any code that has a reference to the array.
+
+[!CODE [csSnippets.Methods#101](../../samples/snippets/csharp/concepts/methods/returnarray1.cs#101)]
+
+## <a name="exten" /> Extension methods ##
+
+Ordinarily, there are two ways to add a method to an existing type:
+
+- Modify the source code for that type. You cannot do this, of course, if you do not own the type's source code. And this becomes a breaking change if you also add any private data fields to support the method.
+- Define the new method in a derived class. A method cannot be added in this way using inheritance for other types, such as structures and enumerations. Nor can it be used to "add" a method to a sealed class.
+
+Extension methods let you "add" a method to an existing type without modifying the type itself or implementing the new method in an inherited type. The extension method also does not have to reside in the same assembly as the type it extends. You call an extension method as if it were a defined member of a type.
+
+For more information, see [Extension Methods](https://msdn.microsoft.com/en-us/library/bb383977.aspx).
+
+## <a name="async" /> Async Methods ##
+
+By using the async feature, you can invoke asynchronous methods without using explicit callbacks or manually splitting your code across multiple methods or lambda expressions.
+
+If you mark a method with the [async](https://msdn.microsoft.com/en-us/library/hh156513.aspx) modifier, you can use the [await](https://msdn.microsoft.com/en-us/library/hh156528.aspx) operator in the method. When control reaches an `await` expression in the async method, control returns to the caller, and progress in the method is suspended until the awaited task completes. When the task is complete, execution can resume in the method.
+
+> [!NOTE]
+> An async method returns to the caller when either it encounters the first awaited object that’s not yet complete or it gets to the end of the async method, whichever occurs first.
+
+An async method can have a return type of @System.Threading.Tasks.Task<TResult>, @System.Threading.Tasks.Task, or `void`. The `void` return type is used primarily to define event handlers, where a `void` return type is required. An async method that returns `void` can't be awaited, and the caller of a void-returning method can't catch exceptions that the method throws.
```

Comment:
I'd leave the sentence as is, and add a note that points to [this page](https://github.com/ljw1004/roslyn/blob/features/async-return/docs/specs/feature%20-%20arbitrary%20async%20returns.md) on the Roslyn repo as a features that is coming in C# 7. We can revisit when the spec gets written.